sourcegraph/sourcegraph
# Code Insights: Migrate repo URLs picker to the MultiCombobox UI

This PR replaces the old repositories URLs picker (which is based on the native text area and plain text for containing repositories URLs) to the new MultiCombobox UI. This PR should improve user interaction with repositories field in the creation UI pages. 

The new MultiCombobox UI field 
- Improves UI jumps, since you have to pick suggestion to add repository we can avoid having additional repository validation (if it's in suggestions that means we have it in our backend)
- Improves UX over repo UI, this might not clear but interactions with textfield might be very complicated, you cursor might be in a wrong position, you can click backspace by accident and this will trigger the live preview, with new UI it's hard to do since we have to explicitly pick one of repos. 

<img width="1113" alt="Screenshot 2023-01-17 at 20 26 26" src="https://user-images.githubusercontent.com/18492575/213034474-c4b17fb3-9e25-429f-a2a4-f3ecebcd0364.png">

## Tech problems with API
MultiCombobox requires exclude API in entities endpoint, you pick repositories this means that we shouldn't have this picked repository in suggestion anymore, unfortunately `repositories` qql query doesn't support exclusions (like excludeIds argument). Usually I would add this to the API but since in code insights we don't use repositories IDs (we rely on repositories URL instead) we don't have any Ids there but just repositories name, so this mean we would have to add `excludeNames` which looks a bit confusing. At the moment we exclude suggestions that were already picked on the client but this mean that we could ended up with no suggestions state, there is no good solution only on the client really. 

## Test plan
- Check that all types of insights creation UI looks good (Search based, Capture group, Compute, Language stats)
- Check that you can create insights (again check it for all insight types)
- Check that edit functionality works properly (three dot context menu on the dashboard page then click edit)
- Check that landing page dynamic insight example works and looks properly 

@leonore (since you're the one who gave a valid feedback about the old fields)



## App preview:

- [Web](https://sg-web-vk-update-repo-field-ui.onrender.com/search)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.
